### PR TITLE
fixed bug in GetDigestionSiteIndices

### PR DESCRIPTION
**Pull request by bintiibrahim**
#1558 in MM
fixed a bug in the code, it wasn't handling multiple preventing cleavages properly.

user had correct protease configuration: **K|,R|,N|[C],N|[F],N|[H],N|[M]**

### CODE REVIEW

File: Proteomics/ProteolyticDigestion/DigestionMotif.cs [hunk 1]
```diff
@@ -127,32 +127,33 @@ public bool Fits(string sequence, int location)
...
                         }
                     }
                 }
 
-                fits = match ? false : true;
+                fits = prevents ? false : true;
             }
 
-            return fits;
+            return (fits, prevents);
```
Comment (rmillikin) [inline Proteomics/ProteolyticDigestion/DigestionMotif.cs]: can you leave the return type as bool and make line 154:

`return fits && !prevents;`

Comment (rmillikin) [reply Proteomics/ProteolyticDigestion/DigestionMotif.cs]: or just get rid of the `prevents` bool entirely and just set `fits = false` on line 139, 146, etc

File: Test/TestDigestionMotif.cs [hunk 2]
```diff
@@ -145,5 +145,28 @@ public static void TestSyntax()
                 Assert.Fail("Exception shold be thrown for incorrect syntax.");
             });
         }
+
+        [Test]
+        public static void TestMultiplePreventingCleavage()
+        {
+            var empty = new List<Modification>();
+            var digestionmotifs = DigestionMotif.ParseDigestionMotifsFromString("N[M]|,N[C]|,H[M]|,H[F]|");
```
Comment (rmillikin) [inline Test/TestDigestionMotif.cs]: `N[M]|` or `N|[M]`? I thought it was the second one? or does it not matter?

Comment (bintiibrahim) [reply Test/TestDigestionMotif.cs]: it doesnt matter

File: Proteomics/ProteolyticDigestion/Protease.cs [hunk 3]
```diff
@@ -172,13 +172,28 @@ internal List<int> GetDigestionSiteIndices(string proteinSequence)
...
+                    var fits = motif.Fits(proteinSequence, r);
+                    if (fits.Item1 && r + motif.CutIndex < proteinSequence.Length)
+                    {
+                        indice = Math.Max(r + motif.CutIndex, indice);
+                    }
+
+                    if (fits.Item2) // if any motif prevents cleave
                     {
-                        indices.Add(r + motif.CutIndex);
+                        preventing.Add(motif);
```
Comment (rmillikin) [inline Proteomics/ProteolyticDigestion/Protease.cs]: I think I see what you're doing here - you're saying that if one motif is prevented, then all motifs are prevented from cleaving at this site. I don't think we want to do that - can you revert all changes to this file?

File: Proteomics/ProteolyticDigestion/DigestionMotif.cs (hunk 1 above)
Comment (rmillikin) [reply Proteomics/ProteolyticDigestion/DigestionMotif.cs]: plz just return one bool - `return fits && !prevents;`

File: Proteomics/ProteolyticDigestion/DigestionMotif.cs [hunk 4]
```diff
@@ -120,39 +120,47 @@ public bool Fits(string sequence, int location)
             // check for inducing cleavage
             for (m = 0; m < InducingCleavage.Length && fits; m++) // handle patterns
             {
-                currentResidue = sequence[location + m];
-                if (!MotifMatches(InducingCleavage[m], currentResidue))
+                if (location + m >= sequence.Length)
```
Comment (rmillikin) [inline Proteomics/ProteolyticDigestion/DigestionMotif.cs]: `for (m = 0; m < InducingCleavage.Length && fits; m++)`
change to 
`for (m = 0; m < InducingCleavage.Length && m < sequence.Length && fits; m++) `
(check my math)

File: Proteomics/ProteolyticDigestion/DigestionMotif.cs [hunk 5]
```diff
@@ -120,39 +120,47 @@ public bool Fits(string sequence, int location)
...
+                    if (location + m + n >= sequence.Length || location - PreventingCleavage.Length + 1 + n < 0)
                     {
-                        match = false;
+                        prevents = false;
                     }
                     else
                     {
-                        currentResidue = CutIndex != 0 ? sequence[location + m + n] : sequence[location - PreventingCleavage.Length + 1 + n];                      
+                        currentResidue = CutIndex != 0 ? sequence[location + m + n] : sequence[location - PreventingCleavage.Length + 1 + n];
                         if (!PreventingCleavage[n].Equals(currentResidue))
```
Comment (rmillikin) [inline Proteomics/ProteolyticDigestion/DigestionMotif.cs]: `if (!PreventingCleavage[n].Equals(currentResidue))`
change to
`if(MotifMatches(PreventingCleavage[n], currentResidue))`

Comment (rmillikin) [reply Proteomics/ProteolyticDigestion/DigestionMotif.cs]: add a unit test for this (wildcard in the preventing cleavage motif)

File: Proteomics/ProteolyticDigestion/Protease.cs [hunk 6]
```diff
@@ -172,13 +172,28 @@ internal List<int> GetDigestionSiteIndices(string proteinSequence)
             var indices = new List<int>();
             for (int r = 0; r < proteinSequence.Length; r++)
             {
+                var indice = -1;
+                var preventing = new List<DigestionMotif>();
```
Comment (rmillikin) [inline Proteomics/ProteolyticDigestion/Protease.cs]: change this to a boolean (cleavage prevented or not)

File: Proteomics/ProteolyticDigestion/Protease.cs [hunk 7]
```diff
@@ -172,13 +172,28 @@ internal List<int> GetDigestionSiteIndices(string proteinSequence)
             var indices = new List<int>();
             for (int r = 0; r < proteinSequence.Length; r++)
             {
+                var indice = -1;
```
Comment (rmillikin) [inline Proteomics/ProteolyticDigestion/Protease.cs]: rename to `cutSiteIndex`

Comment (rmillikin) [reply Proteomics/ProteolyticDigestion/Protease.cs]: add a comment why this is not `r`

File: Proteomics/ProteolyticDigestion/Protease.cs [hunk 8]
```diff
@@ -172,13 +172,28 @@ internal List<int> GetDigestionSiteIndices(string proteinSequence)
...
+                    }
+
+                    if (fits.Item2) // if any motif prevents cleave
                     {
-                        indices.Add(r + motif.CutIndex);
+                        preventing.Add(motif);
                     }
                 }
+
+                // if no motif prevents cleave
```
Comment (rmillikin) [inline Proteomics/ProteolyticDigestion/Protease.cs]:
```
bool anyMotifPrevents = false;
int cutSiteIndex = -1;

foreach(var motif in motifs)
{
var fitsAndPreventsResults = motif.Fits(proteinSequence, r);
bool fits = fitsAndPreventsResults.Item1 && r + motif.CutIndex < proteinSequence.Length;
bool prevents = fitsAndPreventsResults.Item2;

if(prevents)
{
anyMotifPrevents = true;
}

if(fits && !prevents)
{
cutSiteIndex = r + motif.CutIndex;
}
}

if(!anyMotifPrevents && cutSiteIndex >= 0)
{
indices.add(cutSiteIndex);
}
```

**Pull request merged**